# Homework 3 - Classification


# Generic info

## Load Libraries

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split

In [2]:
!ls

sample_data


In [3]:
! rm *.zip
! rm *.csv
! rm *.txt
! rmdir -f sample_data/
! rm bank+marketing.zip.1

rm: cannot remove '*.zip': No such file or directory
rm: cannot remove '*.csv': No such file or directory
rm: cannot remove '*.txt': No such file or directory
rmdir: invalid option -- 'f'
Try 'rmdir --help' for more information.
rm: cannot remove 'bank+marketing.zip.1': No such file or directory


In [4]:
! ls

sample_data


## Load File

In [5]:
! rm *.zip
! rm *.csv
!wget https://archive.ics.uci.edu/static/public/222/bank+marketing.zip

rm: cannot remove '*.zip': No such file or directory
rm: cannot remove '*.csv': No such file or directory
--2024-10-17 06:47:33--  https://archive.ics.uci.edu/static/public/222/bank+marketing.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified
Saving to: ‘bank+marketing.zip’

bank+marketing.zip      [  <=>               ] 999.85K  2.48MB/s    in 0.4s    

2024-10-17 06:47:33 (2.48 MB/s) - ‘bank+marketing.zip’ saved [1023843]



In [6]:
! unzip bank+marketing.zip

Archive:  bank+marketing.zip
 extracting: bank.zip                
 extracting: bank-additional.zip     


In [7]:
! unzip bank.zip

Archive:  bank.zip
  inflating: bank-full.csv           
  inflating: bank-names.txt          
  inflating: bank.csv                


In [8]:
! ls

bank-additional.zip  bank-full.csv	 bank-names.txt  sample_data
bank.csv	     bank+marketing.zip  bank.zip


In [9]:
df = pd.read_csv('bank-full.csv', sep=';')
df_copy = pd.read_csv('bank-full.csv', sep=';')

In [10]:
df_copy.y.value_counts()

y
no     39922
yes     5289
Name: count, dtype: int64

# Question 1

In [11]:
columns = [
    'age',
    'job',
    'marital',
    'education',
    'balance',
    'housing',
    'contact',
    'day',
    'month',
    'duration',
    'campaign',
    'pdays',
    'previous',
    'poutcome',
    'y'
]

In [12]:
df = df[columns]

In [13]:
df.isnull().sum()

age          0
job          0
marital      0
education    0
balance      0
housing      0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64

In [14]:
df.education.mode()

0    secondary
Name: education, dtype: object

Answer = **Secondary**

# Question 2

In [15]:
numerical = df.select_dtypes(include=['float64', 'int64']).columns

In [16]:
numerical

Index(['age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous'], dtype='object')

In [17]:
df[numerical]

age  balance  day  duration  campaign  pdays  previous
0       58     2143    5       261         1     -1         0
1       44       29    5       151         1     -1         0
2       33        2    5        76         1     -1         0
3       47     1506    5        92         1     -1         0
4       33        1    5       198         1     -1         0
...    ...      ...  ...       ...       ...    ...       ...
45206   51      825   17       977         3     -1         0
45207   71     1729   17       456         2     -1         0
45208   72     5715   17      1127         5    184         3
45209   57      668   17       508         4     -1         0
45210   37     2971   17       361         2    188        11

[45211 rows x 7 columns]

In [18]:
df[numerical].corr().round(3)

age  balance    day  duration  campaign  pdays  previous
age       1.000    0.098 -0.009    -0.005     0.005 -0.024     0.001
balance   0.098    1.000  0.005     0.022    -0.015  0.003     0.017
day      -0.009    0.005  1.000    -0.030     0.162 -0.093    -0.052
duration -0.005    0.022 -0.030     1.000    -0.085 -0.002     0.001
campaign  0.005   -0.015  0.162    -0.085     1.000 -0.089    -0.033
pdays    -0.024    0.003 -0.093    -0.002    -0.089  1.000     0.455
previous  0.001    0.017 -0.052     0.001    -0.033  0.455     1.000

Answer = **pdays and previous**

## Encode Y Variable

In [19]:
df.y = (df.y == "yes").astype(int)

In [20]:
df.y.mean()

0.11698480458295547

In [21]:
df.y.value_counts()

y
0    39922
1     5289
Name: count, dtype: int64

In [22]:
df.y.sum()

5289

## Split the data

In [23]:
from sklearn.model_selection import train_test_split

In [24]:
df_train_full, df_test = train_test_split(df, test_size=0.2, random_state=1)

In [25]:
len(df_train_full), len(df_test)

(36168, 9043)

In [26]:
len(df)
len(df) * .2

9042.2

In [27]:
df_train, df_val = train_test_split(df_train_full, test_size=0.25, random_state=11)

In [28]:
len(df_train), len(df_val), len(df_test)

(27126, 9042, 9043)

In [29]:
df_train

age          job   marital  education  balance housing   contact  day  \
36942   37   technician   married  secondary      807     yes  cellular   12   
1511    27      student    single  secondary        0      no   unknown    8   
16776   39   management   married   tertiary       63     yes  cellular   24   
31976   42  blue-collar   married    primary     7735     yes  cellular   13   
15404   48   technician   married   tertiary     2392     yes  cellular   18   
...    ...          ...       ...        ...      ...     ...       ...  ...   
18062   50  blue-collar  divorced    primary     -333      no  cellular   30   
3678    40       admin.   married  secondary        2     yes   unknown   16   
9506    46   technician   married  secondary        0      no   unknown    6   
31658   32   technician    single   tertiary      948     yes  cellular    6   
22988   43   management   married  secondary    17555      no  cellular   26   

      month  duration  campaign  pdays  previous poutcome  y  
36942   may       807         3    172         4  failure  0  
1511    may        93         2     -1         0  unknown  0  
16776   jul       203         1     -1         0  unknown  0  
31976   apr       166         1    144         2  failure  0  
15404   jul       133         5     -1         0  unknown  0  
...     ...       ...       ...    ...       ...      ... ..  
18062   jul       208         2     -1         0  unknown  0  
3678    may        29         1     -1         0  unknown  0  
9506    jun       186         2     -1         0  unknown  0  
31658   apr       223         2     -1         0  unknown  0  
22988   aug        14        14     -1         0  unknown  0  

[27126 rows x 15 columns]

In [30]:
df_train = df_train.reset_index(drop=True)
df_val = df_val.reset_index(drop=True)
df_test = df_test.reset_index(drop=True)

In [31]:
y_train = df_train.y.values
y_val = df_val.y.values
y_test = df_test.y.values

del df_train['y']
del df_val['y']
del df_test['y']

In [32]:
len(df_train), len(df_val), len(df_test)

(27126, 9042, 9043)

In [33]:
len(y_train), len(y_val), len(y_test)

(27126, 9042, 9043)

# Question 3

## Mutual information score

In [34]:
from sklearn.metrics import mutual_info_score

In [35]:
categorical = df.select_dtypes(include=['object', 'category']).columns

In [36]:
categorical

Index(['job', 'marital', 'education', 'housing', 'contact', 'month',
       'poutcome'],
      dtype='object')

In [37]:
df[categorical].dtypes

job          object
marital      object
education    object
housing      object
contact      object
month        object
poutcome     object
dtype: object

In [38]:
y_train

array([0, 0, 0, ..., 0, 0, 0])

In [39]:
categorical = list(df.select_dtypes(include=['object', 'category']).columns)
for c in categorical:
  result = mutual_info_score(y_train, df_train[c])
  print(c, result.round(2))

job 0.01
marital 0.0
education 0.0
housing 0.01
contact 0.01
month 0.03
poutcome 0.03


## Answer = Poutcome

# Question 4

One Hot Encoding

In [40]:
from sklearn.feature_extraction import DictVectorizer

In [41]:
numerical = list(df_train.select_dtypes(include=['float64', 'int64']).columns)
categorical = list(df_train.select_dtypes(include=['object', 'category']).columns)

In [42]:
categorical

['job', 'marital', 'education', 'housing', 'contact', 'month', 'poutcome']

In [43]:
numerical

['age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous']

In [44]:
train_dict = df_train[categorical + numerical].to_dict(orient='records')

In [45]:
train_dict[0]

{'job': 'technician',
 'marital': 'married',
 'education': 'secondary',
 'housing': 'yes',
 'contact': 'cellular',
 'month': 'may',
 'poutcome': 'failure',
 'age': 37,
 'balance': 807,
 'day': 12,
 'duration': 807,
 'campaign': 3,
 'pdays': 172,
 'previous': 4}

In [46]:
dv = DictVectorizer(sparse=False)
dv.fit(train_dict)

DictVectorizer(sparse=False)

In [47]:
X_train = dv.transform(train_dict)

In [48]:
X_train.shape

(27126, 47)

In [50]:
dv.get_feature_names_out()

array(['age', 'balance', 'campaign', 'contact=cellular',
       'contact=telephone', 'contact=unknown', 'day', 'duration',
       'education=primary', 'education=secondary', 'education=tertiary',
       'education=unknown', 'housing=no', 'housing=yes', 'job=admin.',
       'job=blue-collar', 'job=entrepreneur', 'job=housemaid',
       'job=management', 'job=retired', 'job=self-employed',
       'job=services', 'job=student', 'job=technician', 'job=unemployed',
       'job=unknown', 'marital=divorced', 'marital=married',
       'marital=single', 'month=apr', 'month=aug', 'month=dec',
       'month=feb', 'month=jan', 'month=jul', 'month=jun', 'month=mar',
       'month=may', 'month=nov', 'month=oct', 'month=sep', 'pdays',
       'poutcome=failure', 'poutcome=other', 'poutcome=success',
       'poutcome=unknown', 'previous'], dtype=object)

Train and use model


In [51]:
from sklearn.linear_model import LogisticRegression

In [52]:
model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000, random_state=42, solver='liblinear')

In [53]:
val_dict = df_val[categorical + numerical].to_dict(orient='records')
X_val = dv.transform(val_dict)

In [54]:
model.predict_proba(X_val)

array([[0.98882011, 0.01117989],
       [0.96342638, 0.03657362],
       [0.65474515, 0.34525485],
       ...,
       [0.96371255, 0.03628745],
       [0.85745833, 0.14254167],
       [0.22370431, 0.77629569]])

In [55]:
y_pred = model.predict_proba(X_val)[:, 1]

In [56]:
y_pred

array([0.01117989, 0.03657362, 0.34525485, ..., 0.03628745, 0.14254167,
       0.77629569])

In [57]:
deposit = y_pred > 0.5

In [58]:
(y_val == deposit).mean().round(2)

0.9

Answer = **0.9**

# Question 5

Feature elimination

In [60]:
from sklearn.feature_extraction import DictVectorizer
from sklearn.linear_model import LogisticRegression

# Define features
numerical = list(df_train.select_dtypes(include=['float64', 'int64']).columns)
categorical = list(df_train.select_dtypes(include=['object', 'category']).columns)
full_feat = categorical + numerical

# OHE base train split
train_dict = df_train[categorical + numerical].to_dict(orient='records')
dv = DictVectorizer(sparse=False)
dv.fit(train_dict)
X_train = dv.transform(train_dict)

# train base model
model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
model.fit(X_train, y_train)

# OHE base validation split
val_dict = df_val[categorical + numerical].to_dict(orient='records')
X_val = dv.transform(val_dict)

# Prediction
y_pred = model.predict_proba(X_val)[:, 1]
deposit = y_pred > 0.5

# Calculate accuracy
base_accuracy = (y_val == deposit).mean()


# Iterate through feature list
for f in full_feat:

  # Remove feature
  features_reduced = categorical + numerical
  features_reduced.remove(f)

  # One Hot Encoding for train split
  train_dict = df_train[features_reduced].to_dict(orient='records')
  dv = DictVectorizer(sparse=False)
  dv.fit(train_dict)
  X_train = dv.transform(train_dict)

  # Model's Trainning
  model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
  model.fit(X_train, y_train)

  # One Hot Encoding for validation split
  val_dict = df_val[features_reduced].to_dict(orient='records')
  X_val = dv.transform(val_dict)

  # Prediction on validation split
  y_pred = model.predict_proba(X_val)[:, 1]
  deposit = y_pred > 0.5

  # Calculate accuracy
  fe_accuracy = (y_val == deposit).mean()
  delta = base_accuracy - fe_accuracy

  # Print Results

  print('Base Accuracy:', base_accuracy, 'Feature:', f, 'Updated Accuracy:', fe_accuracy, 'Delta:', delta)
  print(features_reduced)
  print('\n')

Base Accuracy: 0.8969254589692546 Feature: job Updated Accuracy: 0.8989161689891617 Delta: -0.001990710019907138
['marital', 'education', 'housing', 'contact', 'month', 'poutcome', 'age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous']


Base Accuracy: 0.8969254589692546 Feature: marital Updated Accuracy: 0.8972572439725724 Delta: -0.00033178500331787486
['job', 'education', 'housing', 'contact', 'month', 'poutcome', 'age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous']


Base Accuracy: 0.8969254589692546 Feature: education Updated Accuracy: 0.8980314089803141 Delta: -0.0011059500110595089
['job', 'marital', 'housing', 'contact', 'month', 'poutcome', 'age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous']


Base Accuracy: 0.8969254589692546 Feature: housing Updated Accuracy: 0.8965936739659367 Delta: 0.00033178500331787486
['job', 'marital', 'education', 'contact', 'month', 'poutcome', 'age', 'balance', 'day', 'duration', 'campaign', 'pdays'

Answer = **age**

* Base Accuracy: 0.8969254589692546 Feature: age Updated Accuracy: 0.8968148639681487 Delta: 0.00011059500110588427
* Base Accuracy: 0.8969254589692546 Feature: balance Updated Accuracy: 0.8972572439725724 Delta: -0.00033178500331787486
* Base Accuracy: 0.8969254589692546 Feature: marital Updated Accuracy: 0.8972572439725724 Delta: -0.00033178500331787486
* Base Accuracy: 0.8969254589692546 Feature: previous Updated Accuracy: 0.8970360539703606 Delta: -0.0001105950011059953




# Question 6

In [63]:
from sklearn.feature_extraction import DictVectorizer
from sklearn.linear_model import LogisticRegression

# Define features
numerical = list(df_train.select_dtypes(include=['float64', 'int64']).columns)
categorical = list(df_train.select_dtypes(include=['object', 'category']).columns)
full_feat = categorical + numerical

# OHE base train split
train_dict = df_train[categorical + numerical].to_dict(orient='records')
dv = DictVectorizer(sparse=False)
dv.fit(train_dict)
X_train = dv.transform(train_dict)

# OHE base validation split
val_dict = df_val[categorical + numerical].to_dict(orient='records')
X_val = dv.transform(val_dict)

c = [0.01, 0.1, 1, 10, 100]

for i in c:
  # train base model
  model = LogisticRegression(solver='liblinear', C=i, max_iter=1000, random_state=42)
  model.fit(X_train, y_train)

  # Prediction
  y_pred = model.predict_proba(X_val)[:, 1]
  deposit = y_pred > 0.5

  # Calculate accuracy
  reg_accuracy = (y_val == deposit).mean().round(3)
  print('C:', i, 'Accuracy:', reg_accuracy)

C: 0.01 Accuracy: 0.895
C: 0.1 Accuracy: 0.897
C: 1 Accuracy: 0.897
C: 10 Accuracy: 0.897
C: 100 Accuracy: 0.897


Answer = **0.1**